### PR TITLE
[red-knot] Enhancing Diagnostics for Compare Expression Inference

### PR DESCRIPTION
**Pull request by cake-monotone**
## Summary

- Refactored comparison type inference functions in `infer.rs`: Changed the return type from `Option` to `Result` to lay the groundwork for providing more detailed diagnostics.
- Updated diagnostic messages.

This is a small step toward improving diagnostics in the future.

Please refer to #13787

## Test Plan

mdtest included!

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -3532,26 +3569,35 @@ fn perform_rich_comparison<'db>(
...
-    let dunder = left.class_member(db, dunder_name);
+    let dunder = left.class_member(db, op.dunder_name());
     if !dunder.is_unbound() {
         // TODO: this currently gives the return type even if the arg types are invalid
         // (e.g. int.__lt__ with string instance should be None, currently bool)
         return dunder
             .call(db, &[Type::Instance(left), Type::Instance(right)])
-            .return_ty(db);
+            .return_ty(db)
+            .ok_or(OperatorUnsupportedError {
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:3588]:
```suggestion
            .ok_or_else(|| OperatorUnsupportedError {
```

**Review by AlexWaygood — COMMENTED** [review]
Thank you! Not at all a full review, just a couple of nits I spotted while skimming

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 1 above)
Comment (cake-monotone) [reply crates/red_knot_python_semantic/src/types/infer.rs]: Thanks for catching those! 👍

**Review by carljm — APPROVED** [review]
Looks great! I just have one minor naming comment; I'll push that and then merge.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -3452,6 +3469,26 @@ impl From<RichCompareOperator> for ast::CmpOp {
...
+            RichCompareOperator::Lt => "__lt__",
+            RichCompareOperator::Le => "__le__",
+            RichCompareOperator::Gt => "__gt__",
+            RichCompareOperator::Ge => "__ge__",
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct OperatorUnsupportedError<'db> {
```
[the remark this answers is not in the document]
Comment (cake-monotone) [reply crates/red_knot_python_semantic/src/types/infer.rs]: Actually, it’s coming from the rule name
( https://github.com/astral-sh/ruff/pull/13819/files#diff-65c2c229c88f4021638c996a7496384000d9e7b53b08426b34e92f120bd30b06R2783)
But you're right, CompareUnsupportedError makes more sense! I think the rule name could be updated as well, but it’s a pretty minor thing.

[the remark this answers is not in the document]
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs]: We could update the rule name, but it's possible we want to share a rule code for all of these (binary, unary, compare operators). Not sure yet, there will be future work on rule categorization.

**Pull request merged**